### DISCUSSION TITLE
Help identifying cause of time-lock

### DISCUSSION THREAD

**Discussion by AstridHornBrorholt** [UPPAALModelChecker/UPPAAL-Meta#171] [Q&A]
Hi :-)
I use many variants of a Bouncing Ball simulation for my experiments, and it seems that they all are plagued by a very rare time-lock that pops up from time to time when using UPPAAL SMC features.

# Version 

Version: `UPPAAL 4.1.20-stratego-10 (rev. 37ECAAA437FE60EE), October 2022`. I think the problem goes as far back as stratego-8. My system is Ubuntu 22.04.1 LTS 64-bit.

# Issue

Sometimes an `E[]()`-query fails with the following message:

![The successors of this state are not well defined. Location Ball(0).RareTimelockHereBall(0).id=0 Ball(0).number_deaths=0 LearnerPlayer.fired=193](https://user-images.githubusercontent.com/11016262/203930023-bdc63a90-c680-4347-98b0-2f9002b8ab6b.png)

It looks to me like it could be a floating point precision issue. Is there a way to switch UPPAAL to use doubles, or another way to mitigate it? 

# Details

I have named the relevant location correspondingly, but I cannot see what in the model could be the cause of it, since I see no invariant that could prevent time from passing in that location:

![A UPPAAL Automaton](https://user-images.githubusercontent.com/11016262/203998243-38dd4a91-fadf-4414-997d-f05674b7f506.png)

I have captured a trace of the time-lock happening, which I have zipped up along with the model here.

[BB_timelock.zip](https://github.com/UPPAALModelChecker/UPPAAL-Meta/files/10093228/BB_timelock.zip)

(Running the query in the model for about 20 minutes will make the time-lock likely to appear.)

The trace I have shows the time-lock in the following state:

![image](https://user-images.githubusercontent.com/11016262/203998578-3314622b-f6b6-4c90-a0c1-25d9b3bc1442.png)

    p = 0.28992628990632385
    v =  -28.894527200873693

The only time progression I can pick is `t = 0.010000000615231326`. I note that by my calculations, this is very close to the time until the ball touches the ground, activating the urgent edge `bounce!`.  ( `p = 0.0004905` )

Trying to find the correct time, it looks to be about `t = 0.010016918`.

**Comment by mikucionisaau** [comment]
Could you please retry your model with Stratego-11 rc1?
It seems that I cannot reproduce it neither with query nor the attached concrete trace. 

> Sometimes an `E[]()`-query fails

I cannot find such query in the attached model, so I just tried the `simulate` with lots of runs and I could not run into timelock issue.

> It looks to me like it could be a floating point precision issue. Is there a way to switch UPPAAL to use doubles, or another way to mitigate it?

UPPAAL already uses `double` in SMC queries to represent `double` and `clock` variable values.

> The only time progression I can pick is `t = 0.010000000615231326`. I note that by my calculations, this is very close to the time until the ball touches the ground, activating the urgent edge `bounce!. ( p = 0.0004905 )`
> Trying to find the correct time, it looks to be about `t = 0.010016918`.

The difference between `t=0.010000000615231326` and `t=0.010016918` is very small so the issue is most likely with precision. 
Notice that the discretization time step (in `Statistical parameters`) is `0.01` so it can easily overshoot the `0.00001` difference and hence miss the opportunity to fire the escape transition, especially when the guard is precise `p==0` and `p` is a clock driven by non-trivial ODE `p'=v && v'==-9.81`. You can also try and tune the integration precision by lowering integration error bounds in `Statistical parameters`, but I am afraid that there isn't enough floating point values to find such `t` where `p==0` exactly.

A more realistic guard condition is `p<=0` or `p<=eps` for some small positive `eps` allowing some interval rather than precise value.

**Reply by mikucionisaau** [reply]
By the way, you can right-click a variable value in the concrete simulator like `p` and add a trajectory plot of its value. 
There is also "JFreeChart" option in the preferences which then allows pan-and-zoom inside the plots.

**Reply by mikucionisaau** [reply]
Now that I am playing with this model, I found several issues:
- In principle, there can be no time-locks in `RareTimelockHere` location as there is no invariant, but I got one run with time-lock message, so the issue persists in Stratego-11-rc1. The time-lock is probably the result of urgent channel interaction, but it should not be there no matter what.
- I wanted to see how far `p` overshoots `0` by asking `E[<=120; 500](min: Ball(0).p)` and it yields `~0` (all samples are 0), it also gives a trace for `Concrete Simulator` where the plot shows that `p` can overshoot as low as `-0.5`.
- Meanwhile the query `simulate[<=120] {Ball(0).p}` yields perfect bounces at `p==0`, so there is an issue in the `Concrete Simulator` integration.
- An alternative method to find overshooting is to use simulate query with predicate: `simulate[<=120;1000000] {Ball(0).p} : 1 : Ball(0).p<0`, which should bring an overshooting trajectory and we could look at it. So far I could not get it, it's very rare.
- Concrete Simulator shows some strange overshooting trace from verifier:
![image](https://user-images.githubusercontent.com/4868894/206411237-284b3316-3479-4c07-867b-e311217f65eb.png)
  First issue is that it shows overshooting, whereas the verifier plots do not show them, which means that the verifier either misses those points, or the concrete simulator re-integrates the trace and shows it somehow differently.
  1 and 2: points are at 0 and below meaning that it had an option to bounce there before overshooting even further.
  3: has a steep vertical rise, which should be impossible as the slope is dictated by the speed, which is never infinite. This artefact could perhaps be explained by point filtering as the picture is zoomed in and those points could have collapsed into single pixel.
- The current time hair in the plot jumps erratically back and forth even though one is stepping in one monotonic direction.

**Reply by mikucionisaau** [reply]
Found the explanation for the weird trajectory in the concrete simulator:
![image](https://user-images.githubusercontent.com/4868894/206415418-db7d465e-8e03-48f2-a4bc-aa81418e8ee3.png)

**Reply by AstridHornBrorholt** [reply]
I have upgrade to Stratego-11 rc1 and I can confirm that the issue persists here.

I'd prefer not to decrease the precision, as my experiment is dependent upon a reasonable precision. I have actually considered upping the precision. 

Changing the guard to `p<=0` does not resolve the issue, as the timelock reappeared after about 10 minutes' runtime. I could not capture a trace of this, as the query `simulate[<=120;1000000] {Ball(0).p} : 1 : Ball(0).p<0` freezes UPPAAL when I try to open the trace.

![image](https://user-images.githubusercontent.com/11016262/206427897-2440846a-90db-48fc-a83f-bb0917ef44be.png)

Using `p <= smallnumber` for `const double smallnumber = 0.000001;` or even `smallnumber = 0.1;` causes a timelock to appear instantly in the same location.

**Reply by mikucionisaau** [reply]
Could you please try the latest Uppaal 5 pre-release (from uppaal.org)?
There were some fixes addressing very similar if not the same issues.

**Reply by AstridHornBrorholt** [reply]
I'm afraid that on `UPPAAL 5.0.0-rc1 (rev. 6E53984EBD0A5178), 2023-03-01` the problem persists, for the ` BB_timelock.xml` file linked in the original post.

It might have gotten less likely. It's hard to say with such a rare occurrence, but I had to run the query 
`simulate[<=120;50000] {1}` 
three times before it appeared.

![image](https://user-images.githubusercontent.com/11016262/222687514-abde2df7-c376-4931-9d52-bc5cb4d6c65d.png)

Pardon the screenshot, but this happens in a state where `x = 0.20000000277`
